File: platforms/ios/AGenUI/Classes/Render/Component/Impl/ImageComponent.swift
//
//  ImageComponent.swift
//  AGenUI
//
// Created on 2026/2/27.
//

import UIKit

/// ImageComponent component implementation (compliant with A2UI v0.9 protocol)
///
/// Supported properties:
/// - url: Image URL string (String)
/// - fit: Scale mode (String: 100%, contain, cover, adapt)
/// - variant: Semantic name (String: icon, avatar, smallFeature, mediumFeature, largeFeature, header)
/// - width: Image width for loading options (String)
/// - height: Image height for loading options (String)
///
/// Design notes:
/// - Uses ObservedImageView (custom UIImageView) with frame change monitoring for transition animations
/// - Integrates with ImageLoaderConfiguration for network image loading with caching
/// - Supports configurable transition animations (default: MagicRevealTransition, 1.5s duration)
/// - Animations only execute when surface.animationEnabled is true
class ImageComponent: Component {
    
    static var defaultTransition: ImageLoadTransition = MagicRevealTransition()
    /// Global default animation duration (seconds)
    static var defaultTransitionDuration: TimeInterval = 1.5
    
    // MARK: - Properties
    
    private var imageView: ObservedImageView?

    // Padding constraints — pin the inner imageView to self with adjustable insets so
    // CSS padding shrinks the image content from borderBox to contentBox.
    private var imageTopConstraint: NSLayoutConstraint?
    private var imageLeadingConstraint: NSLayoutConstraint?
    private var imageTrailingConstraint: NSLayoutConstraint?
    private var imageBottomConstraint: NSLayoutConstraint?
    
    // Track last-applied URL and dimension values to avoid unnecessary reloads
    // when styles is sent in full (key existence ≠ value change).
    private var currentUrl: String = ""
    private var currentWidth: Any?
    private var currentHeight: Any?
    
    // MARK: - Initialization
    
    init(componentId: String, properties: [String: Any]) {
        super.init(componentId: componentId, componentType: "Image", properties: properties)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }


    override func createView(){
        let imageView = ObservedImageView()
        imageView.backgroundColor = .clear
        imageView.contentMode = .scaleAspectFit
        imageView.clipsToBounds = true
        imageView.image = nil
        imageView.translatesAutoresizingMaskIntoConstraints = false

        // Add image view
        addSubview(imageView)
        self.imageView = imageView

        // Pin imageView to self with constants that will be adjusted by CSS padding.
        let topC = imageView.topAnchor.constraint(equalTo: topAnchor, constant: 0)
        let leadingC = imageView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 0)
        let trailingC = imageView.trailingAnchor.constraint(equalTo: trailingAnchor, constant: 0)
        let bottomC = imageView.bottomAnchor.constraint(equalTo: bottomAnchor, constant: 0)
        NSLayoutConstraint.activate([topC, leadingC, trailingC, bottomC])
        self.imageTopConstraint = topC
        self.imageLeadingConstraint = leadingC
        self.imageTrailingConstraint = trailingC
        self.imageBottomConstraint = bottomC
        
        super.createView()
        
    }
    
    // MARK: - Measurement Override
    
    override class func measure(type: String, paramJson: String, maxWidth: Float, widthMode: MeasureMode, maxHeight: Float, heightMode: MeasureMode) -> CGSize {
        var measuredWidth: CGFloat = 0
        var measuredHeight: CGFloat = 0

        if (widthMode == .exactly || widthMode == .atMost) && maxWidth > 0 {
            measuredWidth = CGFloat(maxWidth)
        }

        if (heightMode == .exactly || heightMode == .atMost) && maxHeight > 0 {
            measuredHeight = CGFloat(maxHeight)
        }

        return CGSize(width: measuredWidth, height: measuredHeight)
    }
    
    // MARK: - Component Override
    
    override func updateProperties(_ diff: [String: DiffValue]) {
        super.updateProperties(diff)
        
        if !self.isViewCreated {
            return
        }
        
        // styles is sent in full each time, so key-existence checks are always true.
        // Compare actual values to avoid unnecessary image reloads.
        var urlChanged = false
        var sizeChanged = false
        var yogaWidth: Float = 0
        var yogaHeight: Float = 0
        
        if case .value(let urlValue) = diff["url"] {
            let url = urlValue as? String ?? ""
            urlChanged = (url != currentUrl)
            currentUrl = url
        } else if case .deleted = diff["url"] {
            currentUrl = ""
            urlChanged = true
        }
        
        // Update scale mode (A2UI v0.9 protocol: fit)
        if case .value(let v) = diff["fit"], let fit = v as? String {
            imageView?.contentMode = parseFit(fit)
        } else if case .deleted = diff["fit"] {
            imageView?.contentMode = .scaleAspectFit
        }

        // Apply CSS padding to the inner imageView.
        if case .value(let stylesValue) = diff["styles"], let styles = stylesValue as? [String: Any] {
            applyImagePadding(styles)
            
            let newWidth = styles["width"]
            let newHeight = styles["height"]
            if !dimensionEqual(newWidth, currentWidth) || !dimensionEqual(newHeight, currentHeight) {
                sizeChanged = true
                currentWidth = newWidth
                currentHeight = newHeight
            }
            if let w = newWidth as? NSNumber { yogaWidth = w.floatValue }
            if let h = newHeight as? NSNumber { yogaHeight = h.floatValue }
        }

        // Only reload when URL or size values actually changed
        if urlChanged || (sizeChanged && !currentUrl.isEmpty) {
            loadImage(currentUrl, yogaWidth: yogaWidth, yogaHeight: yogaHeight)
        }
    }
    
    /// Compare two dimension values (Any?) for equality.
    private func dimensionEqual(_ a: Any?, _ b: Any?) -> Bool {
        if a == nil && b == nil { return true }
        if a == nil || b == nil { return false }
        return String(describing: a!) == String(describing: b!)
    }

    private func applyImagePadding(_ styles: [String: Any]) {
        guard let topC = imageTopConstraint,
              let leadingC = imageLeadingConstraint,
              let trailingC = imageTrailingConstraint,
              let bottomC = imageBottomConstraint else { return }

        guard CSSPaddingResolver.hasAnyPaddingKey(styles) else {
            // Reset to zero so prior padding does not stick when removed.
            topC.constant = 0
            leadingC.constant = 0
            trailingC.constant = 0
            bottomC.constant = 0
            return
        }

        let p = CSSPaddingResolver.resolve(styles)
        topC.constant = p.top
        leadingC.constant = p.left
        trailingC.constant = -p.right
        bottomC.constant = -p.bottom
    }
    
    /// Parse scale mode
    /// A2UI v0.9 protocol values: fill, contain, cover, none, scaleDown, 100%, adapt
    ///
    /// CSS object-fit standard:
    ///   none       → no scaling, original size, overflow clipped
    ///   scaleDown → shrink to fit if larger than container, keep original if smaller
    ///
    /// Mapped to match Android ImageView.ScaleType:
    ///   contain    → FIT_CENTER    → .scaleAspectFit
    ///   cover      → CENTER_CROP   → .scaleAspectFill
    ///   fill       → FIT_XY        → .scaleToFill
    ///   none       → CENTER        → .center
    ///   scaleDown → CENTER_INSIDE → .scaleAspectFit (shrink if larger, also enlarge
    ///                if smaller — iOS has no exact CENTER_INSIDE; .scaleAspectFit is
    ///                closest: always shows complete image)
    private func parseFit(_ fit: String) -> UIView.ContentMode {
        switch fit.lowercased() {
        case "fill", "100%":
            return .scaleToFill
        case "contain":
            return .scaleAspectFit
        case "cover":
            return .scaleAspectFill
        case "none":
            // none maps to fill: stretch to fill container, may distort
            return .scaleToFill
        case "scaleDown":
            // CSS object-fit: scaleDown — shrink to fit if larger, keep original if smaller
            return .scaleAspectFit
        case "adapt":
            return .scaleAspectFit
        default:
            return .scaleAspectFit
        }
    }
    
    // MARK: - Private Methods
    
    /// Load image
    private func loadImage(_ src: String, yogaWidth: Float, yogaHeight: Float) {
        guard imageView != nil else { return }
        
        // Cancel previous loading tasks
        cancelCurrentLoadTask()
        
        if src.isEmpty {
            return
        }

        loadNetworkImage(from: src, yogaWidth: yogaWidth, yogaHeight: yogaHeight)
    }
    
    /// Current loading task identifier (for cancellation)
    private var currentTaskId: String?
    
    /// Cancel current image loading task
    private func cancelCurrentLoadTask() {
        guard let taskId = currentTaskId else { return }
        ImageLoaderConfiguration.shared.loader.cancel(for: taskId)
        currentTaskId = nil
    }
    
    /// Load network image (using ImageLoaderConfiguration)
    private func loadNetworkImage(from urlString: String, yogaWidth: Float, yogaHeight: Float) {
        guard let imageView = imageView, let url = URL(string: urlString) else {
            Logger.shared.debug("⚠️ [ImageComponent] Invalid URL or imageView is nil: \(urlString)")
            return
        }
        
        // Clear current taskId (cancellation handled at start of loadImage)
        currentTaskId = nil
        
        // Use globally configured ImageLoader to load image
        // Pass width/height from properties if present, otherwise not
        var options: [String: Any] = [
            ImageLoadOptionsKey.surfaceId: surface?.surfaceId ?? "",
            ImageLoadOptionsKey.componentId: componentId
        ]
        let w = CGFloat(yogaWidth) * Component.BS_POINT_SCALE * UIScreen.main.scale
        let h = CGFloat(yogaHeight) * Component.BS_POINT_SCALE * UIScreen.main.scale
        if w > 0 {
            options[ImageLoadOptionsKey.width] = w
        }
        if h > 0 {
            options[ImageLoadOptionsKey.height] = h
        }
        
        let taskId = ImageLoaderConfiguration.shared.loader.loadImage(from: url, options: options) { [weak self] image, isFromCache, error in
            guard let self else { return }
            
            DispatchQueue.main.async {
                if let image = image {
                    guard let imageView = self.imageView else { return }
                    
                    imageView.image = image
                    
                    // Execute transition animation (only when surface animation is enabled)
                    if self.surface?.animationEnabled == true {
                        imageView.executeTransitionAnimation()
                    }
                } else if let error = error {
                    // Log non-cancel errors for debugging
                    if error as? ImageLoaderError != .cancelled {
                        Logger.shared.warning("[ImageComponent] Failed to load image: \(urlString), componentId: \(self.componentId), error: \(error.localizedDescription)")
                    }
                } else {
                    // No image and no error - unexpected state
                    Logger.shared.warning("[ImageComponent] Image load returned nil without error: \(urlString), componentId: \(self.componentId)")
                }
            }
        }
        
        // Save current taskId
        currentTaskId = taskId
    }
    
    /// ImageView supporting height change monitoring
    class ObservedImageView: UIImageView {
        private var pendingAnimation: Bool = false
        private var isTransitioning: Bool = false
        
        override var frame: CGRect {
            didSet {
                handleFrameChange(frame)
            }
        }
        
        /// Request animation execution (called when image loading completes)
        public func executeTransitionAnimation() {
            if frame.height > 1 && frame.width > 1 {
                performAnimation()
            } else {
                pendingAnimation = true
            }
        }
        
        private func handleFrameChange(_ frame: CGRect) {
            guard frame.height > 1, frame.width > 1, pendingAnimation else { return }
            
            pendingAnimation = false
            performAnimation()
        }
        
        private func performAnimation() {
            guard !isTransitioning else { return }
            
            isTransitioning = true
            ImageComponent.defaultTransition.animate(
                on: self,
                duration: ImageComponent.defaultTransitionDuration
            ) { [weak self] in
                self?.isTransitioning = false
            }
        }
    }
}

File: platforms/android/src/main/java/com/amap/agenui/render/component/impl/ImageComponent.java
package com.amap.agenui.render.component.impl;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.widget.ImageView;

import androidx.annotation.NonNull;

import com.amap.agenui.render.component.A2UIComponent;
import com.amap.agenui.render.image.ImageCallback;
import com.amap.agenui.render.image.ImageLoadOptionsKey;
import com.amap.agenui.render.image.ImageLoadResult;
import com.amap.agenui.render.image.ImageLoaderConfig;
import com.amap.agenui.render.image.ImageLoaderError;
import com.amap.agenui.render.style.StyleHelper;
import com.amap.agenui.render.utils.AGenUILogger;
import com.amap.agenui.render.utils.ImageTransition;
import com.amap.agenui.render.utils.ImageTransitionManager;
import com.amap.agenui.render.utils.ShimmerTransition;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/**
 * Image component implementation - compliant with A2UI v0.9 protocol
 *
 * Supported properties:
 * - url: image URL (DynamicString)
 * - fit: scale mode (contain, cover, fill, none, scaleDown)
 * - variant: size hint (icon, avatar, smallFeature, mediumFeature, largeFeature, header)
 *
 */
public class ImageComponent extends A2UIComponent {

    private static final String TAG = "ImageComponent";

    private Context context;

    private ImageView imageView;

    // Shimmer animation management
    private ShimmerTransition shimmerTransition;
    private ShimmerTransition.ShimmerView currentShimmerView;

    // Currently loading requestId, used for cancellation and stale callback filtering.
    private String currentRequestId;

    // Track last-applied URL and dimension values to avoid unnecessary reloads
    // when styles is sent in full (key existence ≠ value change).
    private String currentUrl = "";
    private Object currentWidth;
    private Object currentHeight;

    public ImageComponent(Context context, String id, Map<String, Object> properties) {
        super(id, "Image");
        if (AGenUILogger.isLoggingEnabled()) {
            AGenUILogger.d(TAG, "[ImageComponent] Constructor called - id: " + id);
            AGenUILogger.d(TAG, "[ImageComponent] Constructor - properties: " + properties);
        }

        // Store only the ApplicationContext to avoid holding a strong reference to Activity and causing leaks
        this.context = context.getApplicationContext();
        if (properties != null) {
            this.properties.putAll(properties);
            AGenUILogger.d(TAG, "[ImageComponent] Constructor - properties saved to base class");
        }
    }

    @Override
    protected View onCreateView(Context context) {
        if (AGenUILogger.isLoggingEnabled()) {
            AGenUILogger.d(TAG, "[ImageComponent] onCreateView called - id: " + getId());
            AGenUILogger.d(TAG, "[ImageComponent] onCreateView - properties: " + properties);
        }

        // Background, border-radius, and border are applied uniformly by StyleHelper through the
        // base class — this component owns only the image content (url, fit, scale type).
        imageView = new ImageView(context);
        imageView.setScaleType(ImageView.ScaleType.CENTER_CROP);

        if (!properties.isEmpty()) {
            AGenUILogger.d(TAG, "[ImageComponent] onCreateView - applying properties immediately");
            onUpdateProperties(properties);
        } else {
            AGenUILogger.w(TAG, "[ImageComponent] onCreateView - properties is EMPTY!");
        }

        return imageView;
    }

    @Override
    protected void onUpdateProperties(Map<String, Object> changedProps) {
        if (imageView == null) {
            if (AGenUILogger.isLoggingEnabled()) {
                AGenUILogger.w(TAG, "onUpdateProperties: imageView is null, id=" + getId());
            }
            return;
        }

        // styles is sent in full each time, so key-existence checks are always true.
        // Compare actual values to avoid unnecessary image reloads.
        boolean urlChanged = false;
        boolean sizeChanged = false;
        float yogaWidth = 0f;
        float yogaHeight = 0f;

        if (changedProps.containsKey("url")) {
            Object urlObj = changedProps.get("url");
            String url = urlObj == null ? "" : extractStringValue(urlObj);
            urlChanged = !url.equals(currentUrl);
            currentUrl = url;
        }

        // Update scale mode (A2UI v0.9 protocol: fit)
        if (changedProps.containsKey("fit")) {
            Object fitObj = changedProps.get("fit");
            String fit = fitObj == null ? "" : String.valueOf(fitObj);
            if (AGenUILogger.isLoggingEnabled()) {
                AGenUILogger.d(TAG, "[ImageComponent] onUpdateProperties - setting fit: " + fit);
            }
            imageView.setScaleType(parseFit(fit));
        }

        // Handle styles (read from styles)
        if (changedProps.containsKey("styles")) {
            @SuppressWarnings("unchecked")
            Map<String, Object> styles = (Map<String, Object>) changedProps.get("styles");
            if (styles != null) {
                StyleHelper.applyCSSPadding(imageView, styles, context);

                Object newWidth = styles.get("width");
                Object newHeight = styles.get("height");
                if (!Objects.equals(newWidth, currentWidth)
                        || !Objects.equals(newHeight, currentHeight)) {
                    sizeChanged = true;
                    currentWidth = newWidth;
                    currentHeight = newHeight;
                }
                if (newWidth instanceof Number) yogaWidth = ((Number) newWidth).floatValue();
                if (newHeight instanceof Number) yogaHeight = ((Number) newHeight).floatValue();
            }
        }

        // Only reload when URL or size values actually changed
        if (urlChanged || (sizeChanged && !currentUrl.isEmpty())) {
            if (AGenUILogger.isLoggingEnabled()) {
                AGenUILogger.d(TAG, "[ImageComponent] onUpdateProperties - loading image url: " + currentUrl);
            }
            loadImage(currentUrl, yogaWidth, yogaHeight);
        }
    }

    /**
     * Load the image.
     * Uses a pluggable ImageLoader to load network images and local resources.
     * Integrates Shimmer loading animation and transition animation.
     */
    private void loadImage(String src, float yogaWidth, float yogaHeight) {
        if (AGenUILogger.isLoggingEnabled()) {
            AGenUILogger.d(TAG, "[ImageComponent] loadImage called - src: " + src);
        }

        if (src == null || src.isEmpty()) {
            AGenUILogger.w(TAG, "[ImageComponent] loadImage - src is null or empty, clearing image");
            imageView.setImageDrawable(null);
            return;
        }

        // Cancel the previous loading task so a recycled component does not receive stale callbacks.
        if (currentRequestId != null) {
            ImageLoaderConfig.getInstance().getLoader().cancel(currentRequestId);
            currentRequestId = null;
        }

        // Start Shimmer animation based on animation toggle
        if (isAnimationEnabled()) {
            if (shimmerTransition == null) {
                shimmerTransition = new ShimmerTransition();
            }
            currentShimmerView = shimmerTransition.startShimmer(imageView);
        }

        // Delegate to the pluggable Loader, build options
        Map<String, Object> options = buildLoadOptions(yogaWidth, yogaHeight);
        final String[] requestIdHolder = new String[1];
        String requestId = ImageLoaderConfig.getInstance()
                .getLoader()
                .loadImage(src, options, new ImageCallback() {
            @Override
            public void onSuccess(@NonNull ImageLoadResult result) {
                String activeRequestId = requestIdHolder[0];
                if (activeRequestId == null || !activeRequestId.equals(currentRequestId)) {
                    return;
                }
                currentRequestId = null;
                imageView.setImageDrawable(result.drawable);
                onImageLoadComplete(result.isFromCache);
            }

            @Override
            public void onFailure(@NonNull ImageLoaderError error) {
                String activeRequestId = requestIdHolder[0];
                if (activeRequestId == null || !activeRequestId.equals(currentRequestId)) {
                    return;
                }
                currentRequestId = null;
                AGenUILogger.e(TAG, "[ImageComponent] Image load failed: " + id + ", error: " + error.getMessage());
                if (shimmerTransition != null) {
                    shimmerTransition.stopShimmer();
                }
                // Set the global default placeholder
                Drawable placeholder = ImageLoaderConfig.getInstance().getDefaultPlaceholder();
                if (placeholder != null) {
                    imageView.setImageDrawable(placeholder);
                }
            }
        });
        requestIdHolder[0] = requestId;
        currentRequestId = requestId;
    }

    /**
     * Build image loading options based on component properties and layout information.
     *
     * <p>When the layout explicitly specifies width and height, pass options to the loader
     * for downsampling optimization.
     *
     * @return options map, or null if no valid options are present
     */
    private Map<String, Object> buildLoadOptions(float yogaWidth, float yogaHeight) {
        Map<String, Object> options = new HashMap<>();

        if (yogaWidth > 0) {
            float px = StyleHelper.standardUnitToPx(context, yogaWidth);
            if (px > 0) options.put(ImageLoadOptionsKey.WIDTH, px);
        }
        if (yogaHeight > 0) {
            float px = StyleHelper.standardUnitToPx(context, yogaHeight);
            if (px > 0) options.put(ImageLoadOptionsKey.HEIGHT, px);
        }

        // Component ID
        options.put(ImageLoadOptionsKey.COMPONENT_ID, getId());

        return options.size() > 1 ? options : null; // don't pass options if only componentId is present
    }

    /**
     * Called when image loading is complete.
     *
     * @param isFromCache whether the image came from cache (skip entrance animation on cache hit)
     */
    private void onImageLoadComplete(boolean isFromCache) {
        // Remove the Shimmer overlay
        if (shimmerTransition != null) {
            shimmerTransition.stopShimmer();
        }

        // Execute reveal animation based on animation toggle (skip on cache hit)
        if (isAnimationEnabled() && !isFromCache) {
            executeTransition();
        }
    }

    /**
     * Cancel the loading task when the component is destroyed.
     */
    public void onDestroy() {
        if (currentRequestId != null) {
            ImageLoaderConfig.getInstance().getLoader().cancel(currentRequestId);
            currentRequestId = null;
        }
    }

    /**
     * Execute the transition animation after image loading completes.
     */
    private void executeTransition() {
        ImageTransition transition = ImageTransitionManager.getDefaultTransition();
        long duration = ImageTransitionManager.getDefaultDuration();

        if (transition != null) {
            transition.animate(imageView, duration, null);
        }
    }

    /**
     * Extract string value (supports DynamicString)
     */
    private String extractStringValue(Object value) {
        if (value instanceof Map) {
            @SuppressWarnings("unchecked")
            Map<String, Object> valueMap = (Map<String, Object>) value;
            if (valueMap.containsKey("literalString")) {
                return String.valueOf(valueMap.get("literalString"));
            }
            if (valueMap.containsKey("path")) {
                return "";
            }
        }
        return String.valueOf(value);
    }

    /**
     * Parse scale mode.
     * A2UI v0.9 protocol values: contain, cover, fill, none, scaleDown
     */
    private ImageView.ScaleType parseFit(String fit) {
        switch (fit.toLowerCase()) {
            case "contain":
                return ImageView.ScaleType.FIT_CENTER;
            case "cover":
                return ImageView.ScaleType.CENTER_CROP;
            case "fill":
                return ImageView.ScaleType.FIT_XY;
            case "none":
                // none maps to fill: stretch to fill container
                return ImageView.ScaleType.FIT_XY;
            case "scaleDown":
                // scaleDown is similar to contain but does not enlarge the image
                return ImageView.ScaleType.CENTER_INSIDE;
            default:
                return ImageView.ScaleType.FIT_CENTER;
        }
    }

}
